#URL   https://www.kaggle.com/datasets/swathiachath/kc-housesales-data

# Predicting Property price 

### Introduction 

##### The dataset belongs to a real estate website in King County, the city in Washington state, USA. They want me to design a machine learning model to predict the price of houses. This dataset consisted of 21 variables and 21613 different houses.The price prediction is an important task for real estate companies because they can give customers much more service with less cost in comparison with the traditional way. Predicting price also has some advantages for the company. Firstly, it can optimize their cost since in the traditional real estate companies the price of a house was not systematic and there were no metrics to calculate error while in machine learning tasks is it. they need much more professional staff than modern companies that use machine learning. Additional to all they can give service to much more customers.They want me to built machine learning pipline base on number of bedrooms and	bathrooms, sqft_living(living square footage), sqftabove (square footage of house apart from basement )sqftbasement (square footage of the basement) ,zipcode, I will explore the data set and formulate it as supervised learning. The target is the price of a house that is why the model should be regression.

## Loding Python Library 

### Loding an library that I will use later 

In [ ]:
import pandas as pd 
import sklearn.model_selection
import sklearn.preprocessing
import sklearn.metrics
import sklearn.decomposition
import imblearn.over_sampling
import sklearn.linear_model
import sklearn.model_selection
import sklearn.ensemble
import sklearn.neural_network
import matplotlib.pyplot as plt
import matplotlib.pyplot as plt
import numpy as np

## Loding Data Set and Exploration :

#URL   https://www.kaggle.com/datasets/swathiachath/kc-housesales-data

In [2]:
df = pd.read_csv("../datasets/kc_house_data.csv")
df = df.set_index("id")
df=df.drop("view",axis=1)

df.head()

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
7129300520  20141013T000000  221900.0         3       1.00         1180   
6414100192  20141209T000000  538000.0         3       2.25         2570   
5631500400  20150225T000000  180000.0         2       1.00          770   
2487200875  20141209T000000  604000.0         4       3.00         1960   
1954400510  20150218T000000  510000.0         3       2.00         1680   

            sqft_lot  floors  waterfront  condition  grade  sqft_above  \
id                                                                       
7129300520      5650     1.0           0          3      7        1180   
6414100192      7242     2.0           0          3      7        2170   
5631500400     10000     1.0           0          3      6         770   
2487200875      5000     1.0           0          5      7        1050   
1954400510      8080     1.0           0          3      8        1680   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
7129300520              0      1955             0    98178  47.5112 -122.257   
6414100192            400      1951          1991    98125  47.7210 -122.319   
5631500400              0      1933             0    98028  47.7379 -122.233   
2487200875            910      1965             0    98136  47.5208 -122.393   
1954400510              0      1987             0    98074  47.6168 -122.045   

            sqft_living15  sqft_lot15  
id                                     
7129300520           1340        5650  
6414100192           1690        7639  
5631500400           2720        8062  
2487200875           1360        5000  
1954400510           1800        7503

#### In this section, I load the data set. Set the id as an index and then the number of views has dropped since the number of views is not a good variable for predicting a house. one house could be viewed many times but it does not mean that it is expencive or not.

In [3]:
print("The shape of dataset is:",df.shape)

The shape of dataset is: (21613, 19)


In [4]:
print("the columns is:",df.columns)

the columns is: Index(['date', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot',
       'floors', 'waterfront', 'condition', 'grade', 'sqft_above',
       'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',
       'sqft_living15', 'sqft_lot15'],
      dtype='object')


####  I will explane the meaning of  varabels that are not clear, sqft_living(living square footage),sqft_lot (square footage of the lot),waterfront which has view ti water front numeric ,condition how good the condition is, sqft(basement square footage of the house apart from basement), sft(basement square footage of the basement), lat(lattitude coordinate numeric), long(longitude coordinate), sqftliving15 (living room area in 2015 ), sqft lot 15 (lot size area in 2015 ), yr_built year of built and yr_renovated year of renovation that later I replace the with the age of house and age of renovation.

### Explicit missing values :

In [5]:
df.isnull().sum()

date             0
price            0
bedrooms         0
bathrooms        0
sqft_living      0
sqft_lot         0
floors           0
waterfront       0
condition        0
grade            0
sqft_above       0
sqft_basement    0
yr_built         0
yr_renovated     0
zipcode          0
lat              0
long             0
sqft_living15    0
sqft_lot15       0
dtype: int64

#### Understood all explicit missing values by isnull.sum method.  All columns are  zero so there are not any explicit missing values in this data set. 

### Implicit missing values :

In [6]:
df["bedrooms"].unique()

array([ 3,  2,  4,  5,  1,  6,  7,  0,  8,  9, 11, 10, 33])

#### It might not possible to house have 33 or 10 or 11 or 0 room 

In [7]:
df.loc[df["bedrooms"] == 33]

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
2402100895  20140625T000000  640000.0        33       1.75         1620   

            sqft_lot  floors  waterfront  condition  grade  sqft_above  \
id                                                                       
2402100895      6000     1.0           0          5      7        1040   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
2402100895            580      1947             0    98103  47.6878 -122.331   

            sqft_living15  sqft_lot15  
id                                     
2402100895           1330        4700

In [8]:
df.loc[df["bedrooms"] == 10]

date      price  bedrooms  bathrooms  sqft_living  \
id                                                                         
627300145   20140814T000000  1148000.0        10       5.25         4590   
5566100170  20141029T000000   650000.0        10       2.00         3610   
8812401450  20141229T000000   660000.0        10       3.00         2920   

            sqft_lot  floors  waterfront  condition  grade  sqft_above  \
id                                                                       
627300145      10920     1.0           0          3      9        2500   
5566100170     11914     2.0           0          4      7        3010   
8812401450      3745     2.0           0          4      7        1860   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
627300145            2090      2008             0    98004  47.5861 -122.113   
5566100170            600      1958             0    98006  47.5705 -122.175   
8812401450           1060      1913             0    98105  47.6635 -122.320   

            sqft_living15  sqft_lot15  
id                                     
627300145            2730       10400  
5566100170           2040       11914  
8812401450           1810        3745

In [9]:
df.loc[df["bedrooms"] == 11]

date     price  bedrooms  bathrooms  sqft_living  \
id                                                                        
1773100755  20140821T000000  520000.0        11        3.0         3000   

            sqft_lot  floors  waterfront  condition  grade  sqft_above  \
id                                                                       
1773100755      4960     2.0           0          3      7        2400   

            sqft_basement  yr_built  yr_renovated  zipcode     lat     long  \
id                                                                            
1773100755            600      1918          1999    98106  47.556 -122.363   

            sqft_living15  sqft_lot15  
id                                     
1773100755           1420        4960

In [10]:
df.loc[df["bedrooms"] == 0]

date      price  bedrooms  bathrooms  sqft_living  \
id                                                                         
6306400140  20140612T000000  1095000.0         0       0.00         3064   
3918400017  20150205T000000   380000.0         0       0.00         1470   
1453602309  20140805T000000   288000.0         0       1.50         1430   
6896300380  20141002T000000   228000.0         0       1.00          390   
2954400190  20140624T000000  1295650.0         0       0.00         4810   
2569500210  20141117T000000   339950.0         0       2.50         2290   
2310060040  20140925T000000   240000.0         0       2.50         1810   
3374500520  20150429T000000   355000.0         0       0.00         2460   
7849202190  20141223T000000   235000.0         0       0.00         1470   
7849202299  20150218T000000   320000.0         0       2.50         1490   
9543000205  20150413T000000   139950.0         0       0.00          844   
1222029077  20141029T000000   265000.0         0       0.75          384   
3980300371  20140926T000000   142000.0         0       0.00          290   

            sqft_lot  floors  waterfront  condition  grade  sqft_above  \
id                                                                       
6306400140      4764     3.5           0          3      7        3064   
3918400017       979     3.0           0          3      8        1470   
1453602309      1650     3.0           0          3      7        1430   
6896300380      5900     1.0           0          2      4         390   
2954400190     28008     2.0           0          3     12        4810   
2569500210      8319     2.0           0          3      8        2290   
2310060040      5669     2.0           0          3      7        1810   
3374500520      8049     2.0           0          3      8        2460   
7849202190      4800     2.0           0          3      7        1470   
7849202299      7111     2.0           0          3      7        1490   
9543000205      4269     1.0           0          4      7         844   
1222029077    213444     1.0           0          3      4         384   
3980300371     20875     1.0           0          1      1         290   

            sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
id                                                                             
6306400140              0      1990             0    98102  47.6362 -122.322   
3918400017              0      2006             0    98133  47.7145 -122.356   
1453602309              0      1999             0    98125  47.7222 -122.290   
6896300380              0      1953             0    98118  47.5260 -122.261   
2954400190              0      1990             0    98053  47.6642 -122.069   
2569500210              0      1985             0    98042  47.3473 -122.151   
2310060040              0      2003             0    98038  47.3493 -122.053   
3374500520              0      1990             0    98031  47.4095 -122.168   
7849202190              0      1996             0    98065  47.5265 -121.828   
7849202299              0      1999             0    98065  47.5261 -121.826   
9543000205              0      1913             0    98001  47.2781 -122.250   
1222029077              0      2003             0    98070  47.4177 -122.491   
3980300371              0      1963             0    98024  47.5308 -121.888   

            sqft_living15  sqft_lot15  
id                                     
6306400140           2360        4000  
3918400017           1470        1399  
1453602309           1430        1650  
6896300380           2170        6000  
2954400190           4740       35061  
2569500210           2500        8751  
2310060040           1810        5685  
3374500520           2520        8050  
7849202190           1060        7200  
7849202299           1500        4675  
9543000205           1380        9600  
1222029077           1920      224341  
3980300371

#### It is possible to have a big property that has 33 or 11 or 10 rooms because the square of the property are big as well. 

In [11]:
df["bathrooms"].unique()

array([1.  , 2.25, 3.  , 2.  , 4.5 , 1.5 , 2.5 , 1.75, 2.75, 3.25, 4.  ,
       3.5 , 0.75, 4.75, 5.  , 4.25, 3.75, 0.  , 1.25, 5.25, 6.  , 0.5 ,
       5.5 , 6.75, 5.75, 8.  , 7.5 , 7.75, 6.25, 6.5 ])

#### The number of bathrooms is weird since the number of bathrooms by itself does not affect the price of the property so I decided to remove it.

In [12]:
df=df.drop("bathrooms",axis=1)

In [13]:
df["sqft_lot"].unique()

array([ 5650,  7242, 10000, ...,  5813,  2388,  1076])

In [14]:
df["sqft_lot"].unique()

array([ 5650,  7242, 10000, ...,  5813,  2388,  1076])

In [15]:
df["floors"].unique()

array([1. , 2. , 1.5, 3. , 2.5, 3.5])

#### It is possible to have a half floor so there is not missing values.

In [16]:
df["condition"].unique()

array([3, 5, 4, 1, 2])

In [17]:
df["grade"].unique()

array([ 7,  6,  8, 11,  9,  5, 10, 12,  4,  3, 13,  1])

In [18]:
df["sqft_above"].unique()

array([1180, 2170,  770, 1050, 1680, 3890, 1715, 1060, 1890, 1860,  860,
       1430, 1370, 1810, 1980, 1600, 1200, 1250, 2330, 2270, 1070, 2450,
       1710, 1750, 1400,  790, 2570, 2320, 1190, 1510, 1090, 1280,  930,
       2360,  890, 2620, 2600, 3595, 1570,  920, 3160,  990, 2290, 2165,
       1640, 1000, 2130, 2830, 2250, 2420, 3250, 1850, 1590, 1260, 2519,
       1540, 1110, 1770, 2720, 2240, 3070, 2380, 2390,  880, 1040,  910,
       3450, 2350, 1900, 1010,  960, 2660, 1610,  765, 3520, 1290, 1960,
       1160, 1210, 1270, 1440, 2190, 2920, 1460, 1170, 1240, 3140, 2030,
       2310,  700, 1080, 2520, 2780, 1560, 1450, 1720, 2910, 1620, 1360,
       2070, 2460, 1390, 2140, 1320, 1340, 1550,  940, 1380, 3670, 2370,
       1130,  980, 3540, 2500, 1760, 1030, 1780, 3400, 2680, 1670, 2590,
        820, 1220, 2440, 2090, 1100, 1330, 1420, 1690, 2150, 1910, 1350,
       1940,  900, 1630, 2714,  850, 1870, 1950, 2760, 2020, 1120, 1480,
       1230, 2280, 3760, 3530,  830, 1300, 2740, 18

In [19]:
df["sqft_basement"].unique()

array([   0,  400,  910, 1530,  730, 1700,  300,  970,  760,  720,  700,
        820,  780,  790,  330, 1620,  360,  588, 1510,  410,  990,  600,
        560,  550, 1000, 1600,  500, 1040,  880, 1010,  240,  265,  290,
        800,  540,  380,  710,  840,  770,  480,  570, 1490,  620, 1250,
       1270,  120,  650,  180, 1130,  450, 1640, 1460, 1020, 1030,  750,
        640, 1070,  490, 1310,  630, 2000,  390,  430,  850,  210, 1430,
       1950,  440,  220, 1160,  860,  580, 2060, 1820, 1180,  200, 1150,
       1200,  680,  530, 1450, 1170, 1080,  960, 1100,  280,  870,  460,
       1400, 1320,  660, 1220,  900,  420, 1580, 1380,  475,  690,  270,
        350,  935, 1370,  980, 1470,  160,  950,   50,  740, 1780, 1900,
        340,  470,  370,  140, 1760,  130,  610,  520,  890, 1110,  150,
       1720,  810,  190, 1290,  670, 1800, 1120, 1810,   60, 1050,  940,
        310,  930, 1390, 1830, 1300,  510, 1330, 1590,  920, 1420, 1240,
       1960, 1560, 2020, 1190, 2110, 1280,  250, 23

In [20]:
df["zipcode"].unique()

array([98178, 98125, 98028, 98136, 98074, 98053, 98003, 98198, 98146,
       98038, 98007, 98115, 98107, 98126, 98019, 98103, 98002, 98133,
       98040, 98092, 98030, 98119, 98112, 98052, 98027, 98117, 98058,
       98001, 98056, 98166, 98023, 98070, 98148, 98105, 98042, 98008,
       98059, 98122, 98144, 98004, 98005, 98034, 98075, 98116, 98010,
       98118, 98199, 98032, 98045, 98102, 98077, 98108, 98168, 98177,
       98065, 98029, 98006, 98109, 98022, 98033, 98155, 98024, 98011,
       98031, 98106, 98072, 98188, 98014, 98055, 98039])

In [21]:
df["yr_built"].unique()

array([1955, 1951, 1933, 1965, 1987, 2001, 1995, 1963, 1960, 2003, 1942,
       1927, 1977, 1900, 1979, 1994, 1916, 1921, 1969, 1947, 1968, 1985,
       1941, 1915, 1909, 1948, 2005, 1929, 1981, 1930, 1904, 1996, 2000,
       1984, 2014, 1922, 1959, 1966, 1953, 1950, 2008, 1991, 1954, 1973,
       1925, 1989, 1972, 1986, 1956, 2002, 1992, 1964, 1952, 1961, 2006,
       1988, 1962, 1939, 1946, 1967, 1975, 1980, 1910, 1983, 1978, 1905,
       1971, 2010, 1945, 1924, 1990, 1914, 1926, 2004, 1923, 2007, 1976,
       1949, 1999, 1901, 1993, 1920, 1997, 1943, 1957, 1940, 1918, 1928,
       1974, 1911, 1936, 1937, 1982, 1908, 1931, 1998, 1913, 2013, 1907,
       1958, 2012, 1912, 2011, 1917, 1932, 1944, 1902, 2009, 1903, 1970,
       2015, 1934, 1938, 1919, 1906, 1935])

In [22]:
df["yr_renovated"].unique()

array([   0, 1991, 2002, 2010, 1999, 1992, 2013, 1994, 1978, 2005, 2008,
       2003, 1984, 1954, 2014, 2011, 1974, 1983, 1945, 1990, 1988, 1957,
       1977, 1981, 1995, 2000, 1998, 1970, 1989, 2004, 1986, 2009, 2007,
       1987, 1973, 2006, 1985, 2001, 1980, 1971, 1979, 1997, 1950, 1969,
       1948, 2015, 1968, 2012, 1963, 1951, 1993, 1962, 1996, 1972, 1953,
       1955, 1982, 1956, 1940, 1976, 1946, 1975, 1958, 1964, 1959, 1960,
       1967, 1965, 1934, 1944])

In [23]:
df["lat"].unique()

array([47.5112, 47.721 , 47.7379, ..., 47.3906, 47.3339, 47.6502])

In [24]:
df["long"].unique()

array([-122.257, -122.319, -122.233, -122.393, -122.045, -122.005,
       -122.327, -122.315, -122.337, -122.031, -122.145, -122.292,
       -122.229, -122.394, -122.375, -121.962, -122.343, -122.21 ,
       -122.306, -122.341, -122.169, -122.166, -122.172, -122.218,
       -122.36 , -122.314, -122.304, -122.11 , -122.07 , -122.357,
       -122.368, -122.157, -122.31 , -122.132, -122.362, -122.282,
       -122.18 , -122.027, -122.347, -122.016, -122.364, -122.175,
       -121.977, -122.371, -122.151, -122.301, -122.451, -122.322,
       -122.189, -122.384, -122.369, -122.281, -122.29 , -122.114,
       -122.122, -122.116, -122.149, -122.339, -122.335, -122.344,
       -122.32 , -122.297, -122.192, -122.215, -122.16 , -122.179,
       -122.287, -122.036, -122.073, -121.987, -122.125, -122.34 ,
       -122.025, -122.008, -122.291, -122.365, -122.199, -122.194,
       -122.387, -122.372, -122.391, -122.351, -122.386, -122.249,
       -122.277, -122.378, -121.958, -121.714, -122.08 , -122.

In [25]:
df["sqft_living15"].unique()

array([1340, 1690, 2720, 1360, 1800, 4760, 2238, 1650, 1780, 2390, 2210,
       1330, 1370, 2140, 1890, 1610, 1060, 1280, 1400, 4110, 2240, 1220,
       2200, 1030, 1760, 1860, 1520, 2630, 2580, 1390, 1460, 1570, 2020,
       1590, 2160, 1730, 1290, 2620, 2470, 2410, 3625, 1580, 3050, 1228,
       2680,  970, 1190, 1990, 1410, 1480, 2730, 1950, 2250, 2690, 2960,
       2270, 2570, 2500, 1440, 2750, 2221, 1010, 3390, 3530, 1640, 1510,
       2420, 1940, 3240, 1680,  890, 1130, 3350, 2350, 1870, 1720, 1850,
       1900, 1980, 2520, 1350, 1750, 1160, 2550, 2370, 1240, 1270, 2990,
       1380, 1540, 2090, 2640, 1830, 1620, 1880, 2340, 1710, 2700, 3060,
       2660, 1700, 1970, 1420, 2060, 2480, 1550, 1170, 2820, 1560, 2230,
       2840, 1450, 1500, 3160, 1200, 3400, 2110, 2920, 1770, 1070, 1930,
       3740, 2260, 1670, 2290, 1050, 2540, 2190, 2030, 1230, 2330, 1300,
       1430, 2770, 1250, 1630, 2590, 2130, 1100, 3836, 1320, 2120, 3070,
       1910, 2080, 1960, 2280, 1150, 3430, 2070, 26

In [26]:
df["sqft_lot15"].unique()

array([5650, 7639, 8062, ..., 5731, 1509, 2007])

### Handling year of built and year of renovation:

In [27]:
df["age_of_house"] = pd.DatetimeIndex(df["date"]).year - df["yr_built"] 
df=df.drop("age_of_house",axis=1)

#### By Date time index I calculate the age of the property. the drop the year of built.

In [28]:
df["age_renovate"] = pd.DatetimeIndex(df["date"]).year - df["yr_renovated"] 
df=df.drop("age_renovate",axis=1)
df=df.drop("date",axis=1)

#### same I do for year of renovate and the drop the year of renovate and date columns as well.

### set the target label:

#### Now is the time to split the feature and target label of the train set and test set. (the axis is equal to one to split by columns)

In [29]:
x = df.drop(["price"],axis=1)
y = df["price"]

In [30]:
x_train,x_test,y_train,y_test = sklearn.model_selection.train_test_split(x,y,train_size=.85, test_size=.15)

#### Split the data into 85% for training set and 15% for testing set.      

### Standardization

In [31]:
scaler = sklearn.preprocessing.StandardScaler()
scaler.fit(x_train)

x_train = scaler.transform(x_train)

x_test= scaler.transform(x_test)


print("x_train_standardized:", x_train.shape)
print("x_test_standardized:", x_test.shape)

x_train_standardized: (18371, 16)
x_test_standardized: (3242, 16)


#### Building and fitting a standardization model and scaling the training and testing features

#### SVD model is one model for reduction regression.first,building a SVD model and fit it and finally try it on training and test set.  The SVD does not gives a good performance. so I deleted this step 

## Model Selection and Hyperparameter Tuning

#### Grid search is used for tunung hyperparameters and set n_job to 10, cv to 5 and the metric is R squer error

### Gradient Boosting 

#### Gradient boosting is a predictive model based on a week prediction ensemble that can be used in regression models because there is no clear way to set hyperparameters and we must do it through trial and error. The number of boosting steps I chose is n_estimators. the learning rate is the rate at which the number of contributions from each tree decreases. max_depth is used to limit the depth of a node in a tree. The proportion of samples to be used for fitting the individual base learning is called a sub-sample. The random seed delivered to each tree estimator at each boosting stage can be controlled by the random stage.

In [32]:
parameters_grid = {'n_estimators':[500,1000,2000],'learning_rate':[.001,0.01,.1],
 'max_depth':[1,2,4],'subsample':[.5,.75,1],'random_state':[1,2]}

model_Gradient= sklearn.model_selection.GridSearchCV(sklearn.ensemble.GradientBoostingRegressor(),parameters_grid,
                                                     n_jobs=10, cv=5,scoring= "r2")

model_Gradient.fit(x_train,y_train)

print("R squer of best  = {:.2f}".format(model_Gradient.best_score_))
print("Best found hyperparameters = {}".format(model_Gradient.best_params_))

R squer of best  = 0.89
Best found hyperparameters = {'learning_rate': 0.1, 'max_depth': 4, 'n_estimators': 1000, 'random_state': 1, 'subsample': 0.5}


### Random forest regression

#### Random forest regression is control over fitting meta estimator.                    n_estimator is the number of trees, the max _feature is for finding the best split, min_sample_split is for splitting an internal node with the fewest amount of samples, and boost_trap sample is used while generating trees.

In [33]:
parameters_grid = {"n_estimators": [10,20,30], "max_features": ["auto", "sqrt", "log2"],
                  
                   "min_samples_split": [2,4,8], "bootstrap": [True, False]
}

model_Random_forest= sklearn.model_selection.GridSearchCV(sklearn.ensemble.RandomForestRegressor(),parameters_grid,
                                                     n_jobs=10, cv=5,scoring= "r2")

model_Random_forest.fit(x_train,y_train)

print(" R squer of bestof best  = {:.2f}".format(model_Random_forest.best_score_))
print("Best found hyperparameters of = {}".format(model_Random_forest.best_params_))

 R squer of bestof best  = 0.87
Best found hyperparameters of = {'bootstrap': False, 'max_features': 'log2', 'min_samples_split': 8, 'n_estimators': 30}


### MLP regression

#### The MLP model is not efficient for this data set because I tried differently hyperparameters but still a lot of time.Also, need more powerful computers for tuning and train them but as we could see Gradient boosting gives much better result with less time. the parameters_grid is the list of parameters I tried.

In [ ]:
parameters_grid = {
    "hidden_layer_sizes":[(100),(150),(100,100),(100,150)],
    "learning_rate":["constant", "invscaling", "adaptive"],
    "batch_size":[2,5,10],
    "max_iter":[21613],
    "solver":["lbfgs", "sgd", "adam"],
    "alpha":[0.00001,0.0001,0.001],
    "random_state":[1,2]
}

model_MLP = sklearn.model_selection.GridSearchCV(sklearn.neural_network.MLPRegressor(),parameters_grid,
                                                     n_jobs=10, cv=5,scoring= "r2")

model_MLP.fit(x_train,y_train)

print("R squer of best of best MLP = {:.2f}".format(model_MLP.best_score_))
print("Best found hyperparameters of MLP = {}".format(model_MLP.best_params_))

## Model assessment:

#### The best model is Gradiant boosting with the parameters that Grid search has chosen. The R square for that model was 89 so in the model assessment section, the y_predict is calculated to understand the efficiency of our model the reason that I chose R square is that it is easy to understand and it means the calculation of the ratio of the dependent variable's variability that the model can account for. Also, I calculated the mean squared error as the second metric, which means calculating the digression of the model prediction from the actual amount.

In [36]:
y_predicted = model_Gradient.predict(x_test)


r2 = sklearn.metrics.r2_score(y_test, y_predicted)


mean_squared_error = sklearn.metrics.mean_squared_error(y_test, y_predicted)


print("R squer =", r2)
print("mean_squared_error =", mean_squared_error)

R squer = 0.8930646670948288
mean_squared_error = 14018116029.025202


## Final discution

#### I found some advantages to this notebook. First the. The gradient boosting model gives good metrics. The mean square of this model is .89 which is the good one for financial machine learning results. In addition to that, this model is deficient since the hyperparameters tuning and training model is not consume a lot of time and computational power. I changed a lot of ti find the best one. However, there are some drawbacks there was no good explanation about collecting some numbers for this data set. For instance, there is no explanation about the reason why bathrooms have continuous numbers. But fortunately, it does not make any problems for predicting. Also, some features need some feature engineering to get used such as the year of building and year of renovation so by date I calculate the age of the property and age of renovation in this way these features became more meaning full. One problem that still is tuning and training the MLP model. I took a lot of time and still does not get the result, this could be the problem related to complexity and lack of computational power in other words, this model is not the good choice for this data set.  I think the bias and variance are balanced enough as a result of the metrics we see. Overall, I think this model and the way the featured engineer and the hyperparameters tune is efficient enough to give a good result and the company now knows how they can trust the prediction of the model which solves their problem.

## Refrence

##### kaggle.com. (n.d.). House_Regression_Project. [online] Available at: https://www.kaggle.com/code/lunamcbride24/house-regression-project [Accessed 20 Jun. 2022].

#### scikit-learn. (n.d.). Gradient Boosting regression. [online] Available at: https://scikit-learn.org/stable/auto_examples/ensemble/plot_gradient_boosting_regression.html.